feat(progress): add domain, test and .gitkeep

EomKun opened the pull request:
##  Summary
- 진행률 도메인 관련 구현

##  Related Issue

##  Changes
- 강좌 진행률, 강의 진행률 도메인 구현과 테스트 코드 제작
- 패키지 구조 .gitkeep을 이용하여 구성

##  Discussion Topic

##  Checklist
- [x] 로컬 테스트 완료
- [ ] lint 통과(로컬)
- [x] 리뷰어 요청(최소 2명), 라벨/스코프 지정
- [ ] 브랜치 최신화(rebase) 및 충돌 없음

EomKun committed:
feat(progress): add domain, test and .gitkeep

- 강좌 진행률, 강의 진행률 도메인 구현과 테스트 코드 제작
- 패키지 구조 .gitkeep을 이용하여 구성

EomKun committed:
feat(progress): add domain, test and .gitkeep

- 강좌 진행률, 강의 진행률 도메인 구현과 테스트 코드 제작
- 패키지 구조 .gitkeep을 이용하여 구성

EomKun committed:
feat(progress): add domain, test and .gitkeep

- 강좌 진행률, 강의 진행률 도메인 구현과 테스트 코드 제작
- 패키지 구조 .gitkeep을 이용하여 구성

youabledev commented on enrollment/src/main/java/com/lxp/enrollment/domain/model/enums/StudyStatus.java:
```diff
@@ -0,0 +1,21 @@
+    IN_PROGRESS("진행 중"),
+    COMPLETED("완료");
+
+    private final String description;
```
description 이 있으니 무엇을 위한 상태 값인지 명확해서 좋네요!
제가 만든 status enum에도 추후 참고해서 추가해 두겠습니다!

EomKun committed:
Merge branch 'develop' into jaewoong/feat/progress

minyong96 commented on enrollment/src/main/java/com/lxp/enrollment/domain/model/CourseStudy.java:
```diff
@@ -0,0 +1,137 @@
+/**
+ * 강좌 진행률 도메인
+ */
+public class CourseStudy {
```
도메인 이벤트트 따로 발행을 안하시나여?

minyong96 replied:
AggregateRoot 추가 해주시면 좋을듯 합니다

EomKun replied:
AggregateRoot 추가 완료했습니다

minyong96 commented on enrollment/src/main/java/com/lxp/enrollment/domain/model/CourseStudy.java:
```diff
@@ -0,0 +1,137 @@
+     * @param lectureStudies 강의 진행률 리스트
+     * @return 생성된 강좌 진행률
+     */
+    public static CourseStudy createCourseStudy(CourseStudyId courseStudyId, List<LectureStudy> lectureStudies) {
```
메서드 이름이 좀 더 간결한게 어떤가요?
CourseStudy.createCourseStudy로 호출되는데 접두사가 중복되는 거 같아여

EomKun replied:
create로 수정 완료했습니다

minyong96 commented on enrollment/src/main/java/com/lxp/enrollment/domain/model/CourseStudy.java:
```diff
@@ -0,0 +1,137 @@
+                .studyStatus(StudyStatus.IN_PROGRESS)
+                .completedAt(null)
+                .lectureStudies(lectureStudies)
+                .build();
```
not null 체크가 따로 없는거 같은데 
lectuerStatuy는 불변인거 같은데 빈 배열로 밖에서 생성되나요?

minyong96 replied:
null 체킹 필요한 부분 한번 확인해주세영

minyong96 replied:
LectureStaudy가 외부에서 생성되는게 나을지 아니면, root에서 생성되는게 나을지 한번 고민한번 해주세용

EomKun replied:
null 체크 완료해서 수정 완료했습니다.
LectureStudy는 course ID 기반으로 리스트 기반으로 찾아올것 같아서 외부에서 생성되어 넣을 것 같습니다.

minyong96 commented on enrollment/src/main/java/com/lxp/enrollment/domain/model/CourseStudy.java:
```diff
@@ -0,0 +1,137 @@
+
+    public List<LectureStudy> getLectureStudies() {
+        return lectureStudies;
+    }
```
getXXX() 보다는 record에서 사용하는거처럼 필드이름따라 메서드명을 짓는건 어떠신가여?

EomKun replied:
getCompletedAt() -> completedAt() 으로 나머지도 전부 수정 완료했습니다.

minyong96 commented on enrollment/src/main/java/com/lxp/enrollment/domain/model/CourseStudy.java:
```diff
@@ -0,0 +1,137 @@
+        this.lectureStudies = builder.lectureStudies;
+    }
+
+    public static class Builder {
```
Builder 클래스는  도메인 엔티티 밖으로 분리하는게 어떨까요? 
도메인 모델은 불변을 지키기 위함인데, 값을 세팅하는 builer가 있는게 조금 어색해서요

EomKun replied:
Builder 클래스를 삭제하고, 생성자 기반으로 생성되도록 수정했습니다.

youabledev commented on enrollment/src/main/java/com/lxp/enrollment/domain/model/LectureStudy.java:
```diff
@@ -0,0 +1,69 @@
+     */
+    public void changeCompleted() {
+        this.isCompleted = true;
+    }
```
CourseStudy에서 LectureStudy 리스트의 completed를 관리하는 코드가 없는 것으로 보입니다!
이럴 경우 LectureStudy의 changeCompleted()가 호출된 뒤 CourseStudy의 recalculateProgress()를 직접 호출해 줘야 CourseStudy에 변경사항이 반영될텐데 의도하신게 맞는지 궁금합니다!

LectureStudy의 완료 여부가 CourseStudy의 totalProgress와 studyStatus에 영향을 미치는  조건이 있어서 이 규칙이 깨지지 않도록 관리되어야 할 것 같습니다!

EomKun replied:
해당 부분 CourseStudy(Root Aggregate)가 LectureStudy를 업데이트 하고 이하 로직을 수행하도록 수정하도록 변경했습니다.
DDD에 아직 미숙하다보니 코드상으로는 몰랐는데 좋은 배움이 되었습니다! 감사합니다!

EomKun committed:
fix(progress): modify domain

- AggregateRoot 추가
- 메서드 이름 축약, 변경
- null 체킹 부분 추가
- Builder 삭제
- test 부분 수정

EomKun committed:
fix(progress): modify domain

- CourseStudyTest 부분 수정
- merge 충돌 처리

minyong96 commented on enrollment/src/main/java/com/lxp/enrollment/domain/model/CourseStudy.java:
```diff
@@ -0,0 +1,96 @@
+     */
+    public void recalculateProgress() {
+        if(this.studyStatus == StudyStatus.COMPLETED) {
+            throw new IllegalStateException("완료 상태의 강의는 진도를 업데이트 할 수 없습니다.");
```
아 재웅님 이거 추후 domain exception으로 정의해주세여

EomKun replied:
넵 확인했습니다

EomKun committed:
fix(progress): modify lecture complete logic

[AS-IS] LectureStudy가 직접 완료를 수정하도록 함
[TO-BE] CourseStudy(Root Aggregate)가 LectureStudy를 업데이트 하고 이하 로직을 수행하도록 수정

EomKun committed:
fix(progress): modify domain

- test 부분 수정
- merge 충돌 부분 해결

youabledev commented on enrollment/src/main/java/com/lxp/enrollment/domain/model/CourseStudy.java:
```diff
@@ -0,0 +1,113 @@
+    /**
+     * 강의 진행상태 업데이트
+     */
+    public void updateLectureProgress(LectureStudyId id) {
```
확인했습니다! 리뷰 반영해 주셔서 감사합니다!
